File: CupCake.cpp
//
// Created by NandM2 on 12/10/2021.
//

#include "CupCake.h"

#include <utility>

CupCake::CupCake(std::string flavor, std::string frosting, std::string sprinkleColor) : Cake((frosting == "cream-cheese" ? 2.15 : 1.95), std::move(flavor), frosting), sprinkleColor(std::move(sprinkleColor)) {}

std::string CupCake::ToString() const {
    return Cake::getFlavor() + " cupcake with " + Cake::getFrosting() + " frosting and " + sprinkleColor + " sprinkles " + BakedGood::ToString();
}

double CupCake::DiscountedPrice(int quantity) {
    if(quantity >= 4) {
        return quantity * (BakedGood::getBasePrice() - 0.40);
    } else if(quantity >= 2) {
        return quantity * (BakedGood::getBasePrice() - 0.30);
    } else {
        return quantity * BakedGood::getBasePrice();
    }
}

File: Dart/bakery_checkout/bin/cupcake.dart
import 'cake.dart';

class CupCake extends Cake {
  final String _sprinkleColor;

  CupCake(String flavor, String frosting, this._sprinkleColor) : super((frosting == "cream-cheese" ? 2.15 : 1.95), flavor, frosting);

  @override
  String toString() {
    return "${super.flavor} cupcake with ${super.frosting} frosting and $_sprinkleColor sprinkles " + super.toString();
  }

  @override
  double discountedPrice(int quantity) {
    if(quantity >= 4) {
      return quantity * (super.basePrice - 0.40);
    } else if(quantity >= 2) {
      return quantity * (super.basePrice - 0.30);
    } else {
      return quantity * super.basePrice;
    }
  }
}
